### PR TITLE
feat(pm): add git protocol spec parsing (1/3)

### PR DESCRIPTION
**Pull request by killagu**
## Summary

Split from #2603 — Part 1 of 3 for git dependency resolution.

- Add `PackageSpec` enum (`Registry`, `Git`, `GitHub`) with `PackageSpec::parse()` associated method
- Add `is_non_registry_spec()` with `#[deprecated]` alias for `is_local_spec`
- Add `ResolveError::Git { url, message }` structured error variant
- Add `GitCloneResult` data type in `traits/git.rs` (data only, no implementation)
- Add `is_http_tarball_spec()` for URL routing

### Architectural decisions
- `PackageSpec::parse()` as associated method (not free function) — Rust idiomatic
- Only `Registry/Git/GitHub` variants — no YAGNI `Http/FileDir/FileTarball`
- `traits/git.rs` contains only data types, not implementations
- Structured `ResolveError::Git { url, message }` instead of `Git(String)`

## Test plan
- [x] `cargo check -p utoo-ruborist` passes
- [x] `cargo test -p utoo-ruborist` — 121 tests + 8 doctests pass
- [ ] Review `PackageSpec::parse()` edge cases

> Part of git dependency resolution: **PR1** → #2603-pr2 → #2603-pr3

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/ruborist/src/model/spec.rs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            }
+        }
+
+        // Default: registry spec — use parse_package_spec for name@version splitting
+        let (name, version_spec) = super::util::parse_package_spec(raw);
+        PackageSpec::Registry {
+            name: name.to_string(),
+            version_spec: version_spec.to_string(),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Fixed. `github:foo` (no `/`) now falls through to `PackageSpec::Git` instead of silently becoming a `Registry` spec. I chose `Git` over `Registry` here because the `github:` prefix is an explicit signal that the user intends non-registry resolution — falling back to registry would be more confusing than a clear git resolution error.

Added test: `test_parse_github_no_slash`.

File: crates/ruborist/src/model/spec.rs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        && (spec.ends_with(".tgz") || spec.ends_with(".tar.gz"))
+}
+
+/// Split a string at the first `#` into (base, optional fragment).
+fn split_fragment(s: &str) -> (&str, Option<&str>) {
+    match s.find('#') {
+        Some(pos) => (&s[..pos], Some(&s[pos + 1..])),
+        None => (s, None),
+    }
+}
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs:276]: Fixed. Rewritten with `split_once('#')` and also added empty-fragment handling — trailing `#` now yields `None` instead of `Some("")`.

File: crates/ruborist/src/resolver/registry.rs [hunk 3]
```diff
@@ -24,6 +24,8 @@ pub enum ResolveError<E> {
     NoVersions(String),
     /// Resolved version not found in manifest
     ManifestNotFound { name: String, version: String },
+    /// Git resolution failed
+    Git(String),
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/resolver/registry.rs]: Fixed. Changed to `Git { url: String, message: String }` with the corresponding `Display` and `Error` impls updated.

File: crates/ruborist/src/model/spec.rs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    (spec.starts_with("https://") || spec.starts_with("http://"))
+        && (spec.ends_with(".tgz") || spec.ends_with(".tar.gz"))
+}
+
+/// Split a string at the first `#` into (base, optional fragment).
+fn split_fragment(s: &str) -> (&str, Option<&str>) {
+    match s.find('#') {
+        Some(pos) => (&s[..pos], Some(&s[pos + 1..])),
+        None => (s, None),
+    }
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs:275]: Fixed. Trailing `#` now yields `commit_ish: None`. Added test `test_parse_git_trailing_hash`.

File: crates/ruborist/src/model/spec.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            }
+        );
+    }
+
+    #[test]
+    fn test_is_http_tarball_spec() {
+        assert!(is_http_tarball_spec("https://example.com/pkg.tgz"));
+        assert!(is_http_tarball_spec("http://example.com/pkg.tar.gz"));
+        assert!(!is_http_tarball_spec("https://example.com/pkg"));
+        assert!(!is_http_tarball_spec("lodash@^4.0.0"));
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Fixed. `is_http_tarball_spec` now strips query/fragment before checking the extension, and tests added for URLs with `?v=1.0` and `#download`.

File: crates/ruborist/src/resolver/registry.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/resolver/registry.rs]: Fixed. Changed to `Git { url: String, message: String }`.

File: crates/ruborist/src/model/spec.rs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        if let Some(rest) = raw.strip_prefix("github:") {
+            let (path, commit_ish) = split_fragment(rest);
+            if let Some((owner, repo)) = path.split_once('/') {
+                return PackageSpec::GitHub {
+                    owner: owner.to_string(),
+                    repo: repo.to_string(),
+                    commit_ish: commit_ish.map(String::from),
+                };
+            }
+        }
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Fixed. `github:foo` (no `/`) now falls through to `PackageSpec::Git` with the full raw string as the URL. Added test `test_parse_github_no_slash`.

File: crates/ruborist/src/model/spec.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs:275]: Fixed. `split_fragment` now returns `None` for empty fragments via `\!frag.is_empty()` guard. Added test `test_parse_git_trailing_hash`.

File: crates/ruborist/src/model/spec.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    }
+}
+
+/// Check if a spec string is an HTTP tarball URL.
+///
+/// Useful for routing specs that look like HTTP tarballs away from
+/// the standard registry/git resolution paths.
+pub fn is_http_tarball_spec(spec: &str) -> bool {
+    (spec.starts_with("https://") || spec.starts_with("http://"))
+        && (spec.ends_with(".tgz") || spec.ends_with(".tar.gz"))
```
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Fixed. Applied the suggested approach — `is_http_tarball_spec` now strips query (`?`) and fragment (`#`) before checking the extension. Tests updated to cover URLs with query parameters and fragments.

File: crates/ruborist/src/model/spec.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Fixed. `github:foo` now falls through to `PackageSpec::Git` instead of `Registry`. Added test `test_parse_github_no_slash`.

File: crates/ruborist/src/resolver/preload.rs [hunk 8]
```diff
@@ -47,19 +47,34 @@ pub struct PreloadStats {
...
+        || spec.starts_with("link:")
+        || spec.starts_with("workspace:")
+        || spec.starts_with("portal:")
+        || spec.starts_with("git+")
+        || spec.starts_with("git://")
+        || spec.starts_with("github:")
+}
+
+/// Check if a spec is a local dependency (file:, link:, workspace:, portal:).
+#[deprecated(note = "use is_non_registry_spec instead")]
```
Comment (elrrrrrrr) [inline crates/ruborist/src/resolver/preload.rs]: 可以直接替换了

File: crates/ruborist/src/model/spec.rs [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+            version_spec: version_spec.to_string(),
+        }
+    }
+}
+
+/// Check if a spec string is an HTTP tarball URL.
+///
+/// Useful for routing specs that look like HTTP tarballs away from
+/// the standard registry/git resolution paths.
+pub fn is_http_tarball_spec(spec: &str) -> bool {
```
Comment (elrrrrrrr) [inline crates/ruborist/src/model/spec.rs]: 新增一个 PackageSpec::Tarball 或者删除 http_tarball_spec 相关逻辑？

File: crates/ruborist/src/lib.rs [hunk 10]
```diff
@@ -76,6 +76,16 @@ pub mod compat {
     pub use crate::model::compatibility::{is_cpu_compatible, is_os_compatible};
 }
 
+/// Package specification types.
+pub mod spec {
+    pub use crate::model::spec::{PackageSpec, is_http_tarball_spec};
```
Comment (elrrrrrrr) [inline crates/ruborist/src/lib.rs]: PackageSpec::Tarball 或者删除对应逻辑

File: crates/ruborist/src/resolver/preload.rs (hunk 8 above)
Comment (killagu) [reply crates/ruborist/src/resolver/preload.rs]: Done. `is_non_registry_spec` 已改为直接委托 `PackageSpec::parse`：

```rust
pub fn is_non_registry_spec(spec: &str) -> bool {
    !matches!(PackageSpec::parse(spec), PackageSpec::Registry { .. })
}
```

同时 `PackageSpec::parse` 补充了 `Local`、`HttpTarball` 和 bare `user/repo` 的处理，确保所有非 registry spec 都不会漏掉。

File: crates/ruborist/src/model/spec.rs (hunk 9 above)
Comment (killagu) [reply crates/ruborist/src/model/spec.rs]: Done. 新增了 `PackageSpec::HttpTarball { url }` variant，`PackageSpec::parse` 中 HTTP tarball URL 现在会被正确解析为 `HttpTarball` 而不是误分类为 `Registry`。`is_http_tarball_spec` 保留作为 `parse` 内部的检测函数，但已从 `lib.rs` 的公共 re-export 中移除。

File: crates/ruborist/src/lib.rs (hunk 10 above)
Comment (killagu) [reply crates/ruborist/src/lib.rs]: Done. 已从 re-export 中移除 `is_http_tarball_spec`，只导出 `PackageSpec`。HTTP tarball 的识别现在通过 `PackageSpec::parse` → `HttpTarball` variant 统一处理。

File: crates/ruborist/src/model/spec.rs [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+    }
+}
+
+/// Error returned when a string has no recognizable protocol prefix.
+#[derive(Debug, Clone, Copy)]
+pub struct ParseProtocolError;
+
+impl fmt::Display for ParseProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "no known protocol prefix")
```
Comment (xusd320) [inline crates/ruborist/src/model/spec.rs]:
```suggestion
        write!(f, "Unsupported protocol prefix")
```

**Comment by xusd320** [comment]
为啥不用 thiserror ？

File: crates/ruborist/src/model/spec.rs [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+            Some((proto, rest)) if proto.is_local() => Self::Local {
+                protocol: proto,
+                path: rest.to_owned(),
+            },
+            Some((Protocol::Http, _)) if has_tarball_extension(raw) => Self::HttpTarball {
+                url: raw.to_owned(),
+            },
+            _ => {
+                // Bare GitHub shorthand: user/repo or user/repo#ref
+                // npm treats "user/repo" (no protocol, not scoped) as github:user/repo
```
Comment (elrrrrrrr) [inline crates/ruborist/src/model/spec.rs:203]:
```
  match Protocol::parse_prefix(raw) {
      Some((Protocol::Git, _))                              => { ... }
      Some((Protocol::GitHub, rest))                        => { ... }
      Some((proto, rest)) if proto.is_local()               => { ... }
      Some((Protocol::Http, _)) if has_tarball_extension(raw) => { ... }  // ← 带 guard
      _ => {
          // 兜底：尝试 bare GitHub shorthand，否则当 registry
          if !raw.starts_with('@') {
              if let Some((owner, repo)) = path.split_once('/') && ... {
                  return Self::GitHub { owner, repo, ... };
              }
          }
          // 默认：registry spec
      }
  }

  关键在第 4 个 arm：Some((Protocol::Http, _)) if has_tarball_extension(raw)

  这个 arm 带了一个 guard（if has_tarball_extension）。在 Rust 的 match 中，pattern 匹配成功但 guard 失败时，不会回去尝试其他 Some 分支，而是直接跳到 _。

  所以当输入是 "https://example.com/pkg"（一个没有 .tgz 后缀的 HTTP URL）时：

  第1步: Protocol::parse_prefix("https://example.com/pkg")
         → Some((Protocol::Http, "example.com/pkg"))  ✓ 匹配到了

  第2步: 逐个 arm 检查
         arm 1: Protocol::Git?     ✗ 是 Http，跳过
         arm 2: Protocol::GitHub?  ✗ 是 Http，跳过
         arm 3: proto.is_local()?  ✗ Http 不是 local，跳过
         arm 4: Protocol::Http?    ✓ 匹配！但 guard 检查 has_tarball_extension("https://example.com/pkg") → false
                guard 失败 → 这个 arm 不走

  第3步: 落入 _ 兜底分支

  第4步: 在兜底分支里
         raw = "https://example.com/pkg"
         raw.starts_with('@')? → false，进入 if
         split_once('/') → Some(("https:", "//example.com/pkg"))
         owner="https:" 非空，repo="//example.com/pkg" 非空
         → 返回 GitHub { owner: "https:", repo: "//example.com/pkg" }  💥 错误！

  一个 HTTP URL 被错误地识别成了 GitHub shorthand。

```

添加一个用例还保证这种场景？

File: crates/ruborist/src/model/spec.rs [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+    (Protocol::Portal, &["portal:"]),
+    (Protocol::Http, &["https://", "http://"]),
+];
+
+impl Protocol {
+    /// Try to detect a protocol prefix from a raw spec string.
+    ///
+    /// Returns the protocol and the remaining string after the matched prefix,
+    /// or `None` if no known protocol prefix is found.
+    pub fn parse_prefix(spec: &str) -> Option<(Self, &str)> {
```
Comment (xusd320) [inline crates/ruborist/src/model/spec.rs]: 函数名语义不太对。parse_prefix 应该只只返回 prefix？ 实际上同时返回了 rest。

**Review by elrrrrrrr — APPROVED** [review]
LGTM

File: crates/ruborist/src/model/spec.rs [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+                Self::Registry {
+                    name: name.to_owned(),
+                    version_spec: version_spec.to_owned(),
+                }
+            }
+        }
+    }
+}
+
+impl FromStr for PackageSpec {
```
Comment (xusd320) [inline crates/ruborist/src/model/spec.rs]: 这个实现有点多余了吧？

**Comment by killagu** [comment]
Addressed @xusd320's review comments:

1. `parse_prefix` → `strip_prefix`：语义上和 `str::strip_prefix` 一致，剥离前缀并返回 `(Protocol, &str)` 余下部分。
2. 移除了 `impl FromStr for PackageSpec`，只保留 `impl From<&str>`。调用方通过 `SpecStr` 扩展 trait 的 `parse_spec()` 使用。
3. 错误信息已在上一轮改为 `"Unsupported protocol prefix"`，并用 `thiserror` derive。

**Review by xusd320 — APPROVED** [review]
